### ISSUE TITLE
Improve color table node format

### ISSUE THREAD

**Issue by lassoan** [Slicer/Slicer#7593]
## Is your feature request related to a problem? Please describe.

Slicer's Colors module handles color tables but it can only store label value, name, color, and opacity. There is a growing need for standardization of specifying content of a segmentation. Terminology support was added to segmentation but its adoption was slow, most likely because it was hard to create and interpret terminology json files.

## Describe the solution you'd like

Improve the color table file format in Slicer:
- Use [standard CSV file format](https://datatracker.ietf.org/doc/html/rfc4180)
- Allow storage of standard terminology codes: to facilitate robust data exchange, DICOM storage, etc.
- Proposed field names:
  - LabelValue
  - Name
  - Color.R, Color.G, Color.B, Color.A
  - Category.CodingSchemeDesignator, Category.CodeValue, Category.CodeMeaning
  - Type.CodingSchemeDesignator, Type.CodeValue, Type.CodeMeaning
  - TypeModifier.CodingSchemeDesignator, TypeModifier.CodeValue, TypeModifier.CodeMeaning
  - AnatomicRegion.CodingSchemeDesignator, AnatomicRegion.CodeValue, AnatomicRegion.CodeMeaning
  - AnatomicRegionModifier.CodingSchemeDesignator, AnatomicRegionModifier.CodeValue, AnatomicRegionModifier.CodeMeaning

## Describe alternatives you've considered

JSON would be a more robust and flexible storage format, but it was proven to be hard for users to edit them. Users are comfortable editing tables in Excel, copy/paste content, etc. while there are no similarly universally accessible editor/viewer for JSON files.

## Additional context

See discussion: https://github.com/lassoan/SlicerMONAIAuto3DSeg/pull/15

### Slicer3 label file format:
(Note that space is used as separator and instead of escaping with quotes, space is replaced by underscore):

```text
#WARNING: This file is automatically generated.
#WARNING: Any edits made to this file may be overwritten.
#
#Filename: Slicer3_2010_Brain_Labels.txt
#Title: Slicer3:2010 Brain Labels
#PageName: Slicer3:2010_Brain_Labels
#Source: https://www.slicer.org/w/index.php/Slicer3:2010_Brain_Labels
#
#Begin table data:
0 Background 0 0 0 0
1 Bone 231 197 108 255
2 Skin 177 122 101 255
3 Muscles 205 92 92 255
4 Fat 238 232 170 255
5 CSF 0 191 255 255
6 Ventricles 65 105 225 255
7 Arteries 250 128 114 255
8 Veins 70 130 180 255
9 Gray_matter 211 211 211 255
10 White_matter 255 255 240 255
11 Tumor 144 238 144 255
12 Edema 224 255 255 255
13 Necrosis 216 191 216 255
14 Structure_1 238 130 238 255
15 Structure_2 255 215 0 255
#EOF 
```

### ITK-Snap label file format:

```text
################################################
# ITK-SnAP Label Description File
# File format: 
# IDX   -R-  -G-  -B-  -A--  VIS MSH  LABEL
# Fields: 
#    IDX:   Zero-based index 
#    -R-:   Red color component (0..255)
#    -G-:   Green color component (0..255)
#    -B-:   Blue color component (0..255)
#    -A-:   Label transparency (0.00 .. 1.00)
#    VIS:   Label visibility (0 or 1)
#    IDX:   Label mesh visibility (0 or 1)
#  LABEL:   Label description 
################################################
    0     0    0    0        0  0  0    "Clear Label"
    1   255    0    0        1  1  1    "Label 1"
    2     0  255    0        1  1  1    "Label 2"
    3     0    0  255        1  1  1    "Label 3"
    4   255  255    0        1  1  1    "Label 4"
    5     0  255  255        1  1  1    "Label 5"
    6   255    0  255        1  1  1    "Label 6"
```

**Comment by pieper** [comment]
Yes, this is a great idea.  Can they be backward compatible or does this need to be a new format?

**Comment by lassoan** [comment]
We would of course be able to read the old files but the current file format has irreparable CSV-incompatibilites: space is used as separator (and therefore space in label name is converted to underscore when writing the file, and underscore is replaced by space when reading the file) and it uses comment lines. Comment lines are not part of the [RFC 4180 standard](https://datatracker.ietf.org/doc/html/rfc4180) and therefore they cause inconvenience and incompatibility issues in many use cases, while comments provides very little value (as by definition its content should be ignored by machine processing).

**Comment by pieper** [comment]
Okay, then we should version it somehow like with a new file extension.  It sounds like this would be a useful format to try to promote for tools like dcmqi, highdicom, ohif/dcmjs, etc too.

**Comment by jcfr** [comment]
Related:
* https://github.com/Slicer/Slicer/issues/6305

**Comment by jcfr** [comment]
> JSON would be a more robust and flexible storage format, but it was proven to be hard for users to edit them.

Submitting the schema to https://www.schemastore.org so that standard editor are aware and offer auto-completion may help.

**Comment by pieper** [comment]
👍 for standardizing and clearly the meaning of the color files.  Perhaps that meaning can have more than one standardized representation (i.e. json and csv - and dicom SEG?) with a well defined lossless conversion.

**Comment by CPBridge** [comment]
In highdicom, this information goes into the Segment Description class:

https://highdicom.readthedocs.io/en/latest/package.html#highdicom.seg.SegmentDescription

Generally the user is responsible for creating a list of these objects using the python API, and then passing this into the Segmentation construction. This is a little tedious and it would make sense to add a higher level way to create this list automatically from some standardized definition file such as the one proposed. I can imagine a small helper function could be added to highdicom to do this, and it may be useful to people.

A couple of things to note:
- In principle DICOM allows multiple anatomic regions per segment, whereas the proposed scheme only allows for one. But this could just be a limitation of the format and I don't think it would be too limiting in practice.
-  There is also the information about the algorithm type and algorithm identification, which this does not address. But perhaps this could be passed into the helper function and assumed to be the same for all segments.
- In addition to the "anatomic regions", there are also the "primary anatomic structures", which it may be good to support: https://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_10.5.html#table_10-7b

**Comment by lassoan** [comment]
Thank you @CPBridge you bring up very good points. For storage of all these additional data that you mentioned we would not need to change the proposed format, just specify some more column names. For now, I would go without these extra columns to keep things simpler. Algorithm type and identification is essential, but probably it is better to specify that outside this table to make the table simpler and easier to reuse between different algorithms.

**Comment by muratmaga** [comment]
> JSON would be a more robust and flexible storage format, but it was proven to be hard for users to edit them. Users are comfortable editing tables in Excel, copy/paste content, etc. while there are no similarly universally accessible editor/viewer for JSON files.

I would encourage JSON. I think the issue is creating the JSON. Perhaps the solution is a small tool that takes a data table in Slicer with relevant fields (can be even copy/pasted from Excel or loaded into Slicer), and then converts it to JSON while doing the scheme checks etc... 

We did create a small tool like that, and since then people had been (almost OK), with terminologies. Maybe a version of this, can be turned into a utility module within Slicer? https://github.com/Imageomics/slicerMorph_JSON_generator

**Comment by muratmaga** [comment]
One more issue, perhaps tangential but related: When the order of the segments are changed (but the terminologies are the same), and one exports a labelmap (because they want to train a ML mode), the assigned label index changes between in the resultant labelmap. This is a confusing behavior. If I called my segments bone, tissue, I would expect the indices assigned them to be unique and consistent whether the order is tissue, bone, skin, or skin, tissue, bone. 

I understand the ordering matter for flattening the segmentation, but the indices I believe remain the same. 

When your updating color table / segmentation terminologies, it would be consider this too as well.

**Comment by muratmaga** [comment]
> The most important application of color tables is to prevent this. Voxel values in the exported segmentation are independent of the order of segments if a color table is selected for the export.

But as it stands, I have to specify the color table at the export time explicitly, right? To me this just seems awkward, particularly if I chose to name my segments using a standardized terminology and coloring. What I am trying to get at, the part of that terminology DB should have field like "LabelMap Export Index" or something that should define consistently what the index number of that named segment should be, without having to specify a color table. I may not be seeing the whole picture here, but this seems like a reasonable expectation on user end. Color table is great, if I need change that indices, or there are gaps in sequence etc...

> The format is not just for Slicer but for all AI segmentation tools and applications, so we would need tools on several platforms.

If the format is public, I am not sure why the tool support needs to be on many different platform, if I our goal is to increase the adoption of usage of Terminologies in Slicer. But, anyways, if the tool need to support different communities, it can be a web based application. It shouldn't be difficult to turn what we have to a web based front end, I can ask the people who help developed for us for idea. 

I still think the issue is not the tool, but what the format will contain, particularly if it is meant to serve the whole clinical imaging community.

**Comment by lassoan** [comment]
We discussed this further with @cpinter. Summary:
- Colors module will depend on Terminologies module (Terminologies will provide "presets" for color table terminology fields)
- Colors module will have a single extra "Terminology" column. The displayed content will be set from the terminology with the same method that generates "Name" inputbox in the Terminology navigator widget (e.g., "left kidney", "needle in "needle in Kidney, Left").
- If the user clicks on a Terminology field then a terminology editor widget will be shown. This widget will have 5 collapsible sections (category, type, type modifier, region, region modifier), each with 3 text inputboxes (coding scheme designator, code value, code meaning). The widget could also have a button "Select from terminology", which would show the existing Terminology Navigator widget, and if the user selects a terminology then it would populate all the fields.

In the long term, it may be possible to combine the terminology navigator and terminology editor. For bulk editing of terminologies Excel can be used, too, as the color table files are simple CSV files.

**Comment by cpinter** [comment]
Can we close this ticket?

**Comment by lassoan** [comment]
Yes, this can be closed. For reference, the documentation of the implemented color table format is available here:

https://slicer.readthedocs.io/en/latest/developer_guide/modules/colors.html#color-table-csv-file-format-csv
